In [ ]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
#for dirname, _, filenames in os.walk('/kaggle/input'):
  #  for filename in filenames:
   #     print(os.path.join(dirname, filename))

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Importing Liberaries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from pandas import read_csv, set_option
from pandas.plotting import scatter_matrix
from sklearn.metrics import confusion_matrix

In [ ]:
# Loading Data
fraud_data = pd.read_csv('/content/drive/MyDrive/archive (2)/creditcard.csv')

In [ ]:
# Viewing Raw Data
set_option('display.width', 100)
fraud_data.head()

Time        V1        V2        V3        V4        V5        V6        V7        V8        V9  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599  0.098698  0.363787   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803  0.085102 -0.255425   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461  0.247676 -1.514654   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609  0.377436 -1.387024   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941 -0.270533  0.817739   

   ...       V21       V22       V23       V24       V25       V26       V27       V28  Amount  \
0  ... -0.018307  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   

   Class  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 31 columns]

In [ ]:
# Dimension of data
fraud_data.shape

(284807, 31)

In [ ]:
# Data Type
fraud_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

> Our observations are as 

> NaN values do not present in the data set. Because of the Non-Null Count and number of rows in the dataset match.

> There are 29 Input Variables and 1 Output Variable (Class)

> The data type of all the input variables is float64 whereas the data type of out variable (Class) is int64

In [ ]:
# CHecking Null values
fraud_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
# Summarizing data

fraud_data = fraud_data.round(5)
fraud_data.describe()


Time            V1            V2            V3            V4            V5  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575 -1.158679e-09 -1.509794e-09 -7.794752e-09 -6.390292e-09  9.269435e-09   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00  1.380247e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683170e+00 -1.137433e+02   
25%     54201.500000 -9.203700e-01 -5.985500e-01 -8.903650e-01 -8.486400e-01 -6.915950e-01   
50%     84692.000000  1.811000e-02  6.549000e-02  1.798500e-01 -1.985000e-02 -5.434000e-02   
75%    139320.500000  1.315645e+00  8.037250e-01  1.027195e+00  7.433400e-01  6.119300e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382560e+00  1.687534e+01  3.480167e+01   

                 V6            V7            V8            V9  ...           V21           V22  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  ...  2.848070e+05  2.848070e+05   
mean   9.550327e-09  7.900087e-09 -3.054700e-09 -3.862265e-10  ... -9.550327e-09  1.439571e-09   
std    1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00  ...  7.345240e-01  7.257016e-01   
min   -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01  ... -3.483038e+01 -1.093314e+01   
25%   -7.683000e-01 -5.540750e-01 -2.086300e-01 -6.431000e-01  ... -2.283950e-01 -5.423500e-01   
50%   -2.741900e-01  4.010000e-02  2.236000e-02 -5.143000e-02  ... -2.945000e-02  6.780000e-03   
75%    3.985650e-01  5.704400e-01  3.273500e-01  5.971400e-01  ...  1.863750e-01  5.285550e-01   
max    7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01  ...  2.720284e+01  1.050309e+01   

                V23           V24           V25           V26           V27           V28  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   1.000678e-08  7.689418e-09 -4.915609e-10  4.810275e-09  5.933843e-09 -1.674818e-08   
std    6.244603e-01  6.056471e-01  5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01   
min   -4.480774e+01 -2.836630e+00 -1.029540e+01 -2.604550e+00 -2.256568e+01 -1.543008e+01   
25%   -1.618500e-01 -3.545850e-01 -3.171450e-01 -3.269850e-01 -7.084000e-02 -5.296000e-02   
50%   -1.119000e-02  4.098000e-02  1.659000e-02 -5.214000e-02  1.340000e-03  1.124000e-02   
75%    1.476400e-01  4.395300e-01  3.507200e-01  2.409550e-01  9.104500e-02  7.828000e-02   
max    2.252841e+01  4.584550e+00  7.519590e+00  3.517350e+00  3.161220e+01  3.384781e+01   

              Amount          Class  
count  284807.000000  284807.000000  
mean       88.349619       0.001727  
std       250.120109       0.041527  
min         0.000000       0.000000  
25%         5.600000       0.000000  
50%        22.000000       0.000000  
75%        77.165000       0.000000  
max     25691.160000       1.000000  

[8 rows x 31 columns]

In [ ]:
# Response Variable Analysis
class_names = {0:'Not Fraud', 1:'Fraud'}
rvs = fraud_data.Class.value_counts().rename(index = class_names)
print(rvs)

Not Fraud    284315
Fraud           492
Name: Class, dtype: int64


***Splitting data into training and testing data***

In [ ]:
from sklearn.model_selection import train_test_split

y= fraud_data["Class"]
X = fraud_data.loc[:, fraud_data.columns != 'Class']
X_train,X_test,y_train,y_test = train_test_split(X,y, test_size=1/6, random_state=42)

# Data Modelling

***Logistic Regression***

In [ ]:
#Import Library for Accuracy Score
from sklearn.metrics import accuracy_score

#Import Library for Logistic Regression
from sklearn.linear_model import LogisticRegression

#Initialize the Logistic Regression Classifier
logisreg = LogisticRegression()

#Train the model using Training Dataset
logisreg.fit(X_train, y_train)

# Prediction using test data
y_pred = logisreg.predict(X_test)

# Calculate Model accuracy by comparing y_test and y_pred
acc_logisreg = round( accuracy_score(y_test, y_pred) * 100, 2 )
print( 'Accuracy of Logistic Regression model : ', acc_logisreg )

Accuracy of Logistic Regression model :  99.89


/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


***Linear Discriminent Analysis***

In [ ]:
#Import Library for Linear Discriminant Analysis
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

#Initialize the Linear Discriminant Analysis Classifier
model = LinearDiscriminantAnalysis()

#Train the model using Training Dataset
model.fit(X_train, y_train)

# Prediction using test data
y_pred = model.predict(X_test)

# Calculate Model accuracy by comparing y_test and y_pred
acc_lda = round( accuracy_score(y_test, y_pred) * 100, 2 )
print( 'Accuracy of Linear Discriminant Analysis Classifier: ', acc_lda )

Accuracy of Linear Discriminant Analysis Classifier:  99.93


***Gaussian Naive Bayes***

In [ ]:
#Import Library for Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB

#Initialize the Gaussian Naive Bayes Classifier
model = GaussianNB()

#Train the model using Training Dataset
model.fit(X_train, y_train)

# Prediction using test data
y_pred = model.predict(X_test)

# Calculate Model accuracy by comparing y_test and y_pred
acc_ganb = round( accuracy_score(y_test, y_pred) * 100, 2 )
print( 'Accuracy of Gaussian Naive Bayes : ', acc_ganb )

Accuracy of Gaussian Naive Bayes :  99.3


***Decision Tree***

In [ ]:
#Import Library for Decision Tree Classifier
from sklearn.tree import DecisionTreeClassifier

#Initialize the Decision Tree Classifier
model = DecisionTreeClassifier()

#Train the model using Training Dataset
model.fit(X_train, y_train)

# Prediction using test data
y_pred = model.predict(X_test)

# Calculate Model accuracy by comparing y_test and y_pred
acc_dtree = round( accuracy_score(y_test, y_pred) * 100, 2 )
print( 'Accuracy of  Decision Tree Classifier : ', acc_dtree )

Accuracy of  Decision Tree Classifier :  99.91


***Random Forest***

In [ ]:
#Import Library for Random Forest
from sklearn.ensemble import RandomForestClassifier

#Initialize the Random Forest
model = RandomForestClassifier()

#Train the model using Training Dataset
model.fit(X_train, y_train)

# Prediction using test data
y_pred = model.predict(X_test)

# Calculate Model accuracy by comparing y_test and y_pred
acc_rf = round( accuracy_score(y_test, y_pred) * 100, 2 )
print( 'Accuracy of  Random Forest : ', acc_rf )

Accuracy of  Random Forest :  99.96


***Support Vector Machine***

In [ ]:
#Import Library for Support Vector Machine
from sklearn import svm

#Initialize the Support Vector Classifier
model = svm.SVC()

#Train the model using Training Dataset
model.fit(X_train, y_train)

# Prediction using test data
y_pred = model.predict(X_test)

# Calculate Model accuracy by comparing y_test and y_pred
acc_svc = round( accuracy_score(y_test, y_pred) * 100, 2 )
print( 'Accuracy of Support Vector Classifier: ', acc_svc )

Accuracy of Support Vector Classifier:  99.83


***KNN***

In [ ]:
#Import Library for K Nearest Neighbour Model
from sklearn.neighbors import KNeighborsClassifier

#Initialize the K Nearest Neighbour Model with Default Value of K=5
model = KNeighborsClassifier()

#Train the model using Training Dataset
model.fit(X_train, y_train)

# Prediction using test data
y_pred = model.predict(X_test)

# Calculate Model accuracy by comparing y_test and y_pred
acc_knn = round( accuracy_score(y_test, y_pred) * 100, 2 )
print( 'Accuracy of KNN Classifier: ', acc_knn )

Accuracy of KNN Classifier:  99.84


# Model Selection

In [ ]:
models = pd.DataFrame({
    'Model': ['Logistic Regression', 'Linear Discriminant Analysis','Naive Bayes', 'Decision Tree', 'Random Forest', 'Support Vector Machines', 
              'K - Nearest Neighbors'],
    'Score': [acc_logisreg, acc_lda, acc_ganb, acc_dtree, acc_rf, acc_svc, acc_knn]})

models.sort_values(by='Score', ascending=False)


Model  Score
4                 Random Forest  99.96
1  Linear Discriminant Analysis  99.93
3                 Decision Tree  99.91
0           Logistic Regression  99.89
6         K - Nearest Neighbors  99.84
5       Support Vector Machines  99.83
2                   Naive Bayes  99.30

>The Best model for Predicting in **Random Forest Model** with **99.96% Accuracy**